File: backend/internal/handlers/races.go
package handlers

import (
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	"github.com/keweenaw-endurance/backend/internal/models"
)

type createRaceRequest struct {
	EventID         string  `json:"event_id" binding:"required"`
	Name            string  `json:"name" binding:"required"`
	RaceType        string  `json:"race_type" binding:"required"`
	DistanceKm      float64 `json:"distance_km"`
	DurationMinutes int     `json:"duration_minutes"`
	StartTime       string  `json:"start_time"`
	Status          string  `json:"status"`
}

func (h *Handlers) GetRaces(c *gin.Context) {
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "20"))

	var eventID *uuid.UUID
	if eventIDStr := c.Query("event_id"); eventIDStr != "" {
		id, err := parseUUID(eventIDStr)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "invalid event_id"})
			return
		}
		eventID = &id
	}

	races, total, err := h.services.Races.ListRaces(page, limit, eventID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to list races"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"data":  races,
		"total": total,
		"page":  page,
		"limit": limit,
	})
}

func (h *Handlers) CreateRace(c *gin.Context) {
	var req createRaceRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	eventID, err := parseUUID(req.EventID)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid event_id"})
		return
	}

	race := &models.Race{
		EventID:         eventID,
		Name:            req.Name,
		RaceType:        req.RaceType,
		DistanceKm:      req.DistanceKm,
		DurationMinutes: req.DurationMinutes,
		Status:          req.Status,
	}
	if req.StartTime != "" {
		startTime, err := parseTimestamp(req.StartTime)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "invalid start_time format, use RFC3339"})
			return
		}
		race.StartTime = startTime
	}

	created, err := h.services.Races.CreateRace(race)
	if err != nil {
		respondServiceError(c, err)
		return
	}

	c.JSON(http.StatusCreated, created)
}

func (h *Handlers) GetRace(c *gin.Context) {
	id, err := parseUUID(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid race id"})
		return
	}

	race, err := h.services.Races.GetRace(id)
	if err != nil {
		respondServiceError(c, err)
		return
	}

	c.JSON(http.StatusOK, race)
}

func (h *Handlers) UpdateRace(c *gin.Context) {
	id, err := parseUUID(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid race id"})
		return
	}

	var req createRaceRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	update := &models.Race{
		Name:            req.Name,
		RaceType:        req.RaceType,
		DistanceKm:      req.DistanceKm,
		DurationMinutes: req.DurationMinutes,
		Status:          req.Status,
	}
	if req.StartTime != "" {
		startTime, err := parseTimestamp(req.StartTime)
		if err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "invalid start_time format, use RFC3339"})
			return
		}
		update.StartTime = startTime
	}

	race, err := h.services.Races.UpdateRace(id, update)
	if err != nil {
		respondServiceError(c, err)
		return
	}

	c.JSON(http.StatusOK, race)
}

func (h *Handlers) DeleteRace(c *gin.Context) {
	id, err := parseUUID(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid race id"})
		return
	}

	if err := h.services.Races.DeleteRace(id); err != nil {
		respondServiceError(c, err)
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "race deleted"})
}

func parseDate(value string) (time.Time, error) {
	return time.Parse("2006-01-02", value)
}

func parseTimestamp(value string) (time.Time, error) {
	return time.Parse(time.RFC3339, value)
}

File: frontend/src/stores/races.js
import { defineStore } from 'pinia'
import { racesApi } from '../services/api.js'

export const useRacesStore = defineStore('races', {
  state: () => ({
    races: [],
    currentRace: null,
    total: 0,
    page: 1,
    limit: 20,
    loading: false,
    error: null,
  }),

  actions: {
    async fetchRaces(params = {}) {
      this.loading = true
      this.error = null
      try {
        const { data } = await racesApi.list(params)
        this.races = data.data ?? []
        this.total = data.total ?? 0
        this.page = data.page ?? this.page
        this.limit = data.limit ?? this.limit
      } catch (err) {
        this.error = err.message ?? 'Failed to fetch races'
      } finally {
        this.loading = false
      }
    },

    async fetchRace(id) {
      this.loading = true
      this.error = null
      try {
        const { data } = await racesApi.get(id)
        this.currentRace = data
      } catch (err) {
        this.error = err.message ?? 'Failed to fetch race'
      } finally {
        this.loading = false
      }
    },

    async createRace(payload) {
      const { data } = await racesApi.create(payload)
      this.races.push(data)
      this.total += 1
      return data
    },

    async updateRace(id, payload) {
      const { data } = await racesApi.update(id, payload)
      const index = this.races.findIndex((r) => r.id === id)
      if (index !== -1) {
        this.races[index] = data
      }
      if (this.currentRace?.id === id) {
        this.currentRace = data
      }
      return data
    },

    async deleteRace(id) {
      await racesApi.remove(id)
      this.races = this.races.filter((r) => r.id !== id)
      this.total = Math.max(0, this.total - 1)
      if (this.currentRace?.id === id) {
        this.currentRace = null
      }
    },
  },
})
